support-services/src/main/scala/com/gu/support/redemption/CodeValidationResult.scala

```diff
@@ -0,0 +1,30 @@
+package com.gu.support.redemption
+
+import com.gu.support.redemption.corporate.CorporateCodeValidator.CorporateId
+
+sealed abstract class CodeValidationResult(val clientCode: String)
+
+sealed abstract class ValidCode(clientCode: String) extends CodeValidationResult(clientCode)
+
+sealed abstract class InvalidCode(clientCode: String) extends CodeValidationResult(clientCode)
+
+case object CodeNotFound extends CodeValidationResult("code_not_found")
```

Comment:
why is "code not found" not invalid? (in terms of the ADT it isn't)

Reply:
Because it only means that the code was not found in one of the two systems not that we have determined that there is a specific problem with it. I can see how you could decide to make it extend `InvalidCode` but I think this approach is more flexible

Reply:
yeah makes sense, I agree with your point about not subclassing, based on that I am wondering in my head whether the naming could be better, or we could call it a CorporateCodeNotFound to contain the intermediate state, and have a separate CodeNotFound that covers eventual failure and extends InvalidCode.
However that may get messy in pursuit of trying to reduce the WTF factor?

Reply:
I don't think it's right to say that there's not much shared between the two code types
If we look at the concrete types:
```
case object CodeNotFound extends CodeStatus("code_not_found")
case object CodeMalformed extends InvalidCode("code_malformed")
case object CodeAlreadyUsed extends InvalidCode("code_already_used")
case object CodeExpired extends InvalidCode("code_expired")
case object InvalidReaderType extends InvalidCode("invalid_reader_type")
```
They could all apply to both types of redemption. We don't currently check expiry for corporate codes but I actually think that's an oversight and we will probably need to add it. Similarly `InvalidReaderType` is a problem which could occur in either case.

Reply:
fair point yes you are right there.  I still don't know why I have unease about this type so I don't think it should stop it going into prod!